### PR TITLE
Adding bm3d

### PR DESCRIPTION
**Pull request by bwohlberg**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by ymirm** [comment]
I confirm I am willing to be listed as a maintainer.

File: recipes/bm3d/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/bm3d-{{ version }}.tar.gz
+  sha256: ca6e416f9ec491ca3c408ad796efd3115479177998d50b2c1a2f83fda75648ef
+
+build:
+  number: 0
+  noarch: python
```
Comment (ryanvolz) [inline recipes/bm3d/meta.yaml:15]: For starters, this can't be a `noarch: python` package if you're building and including C extensions for Python.

**Review by ryanvolz — COMMENTED** [review]
Judging from a quick glance at the Linux error messages, it looks like you'll need at least some sort of blas host dependency. See https://conda-forge.org/docs/maintainer/knowledge_base.html#blas for some pointers.

File: recipes/bm3d/meta.yaml (hunk 1 above)
Comment (bwohlberg) [reply recipes/bm3d/meta.yaml]: Thanks; removed it.

**Comment by bwohlberg** [comment]
> Judging from a quick glance at the Linux error messages, it looks like you'll need at least some sort of blas host dependency. See https://conda-forge.org/docs/maintainer/knowledge_base.html#blas for some pointers.

Thanks for the help. I've already tried to address the blas-related error, but it's not clear which of the `openblas` and `libopenblas` packages is the appropriate dependency here (the differences between them doesn't seem to be documented anywhere), and what is particularly strange is that the way I had specified this dependency seemed sufficient for the build phase, but was failing in the test phase. 

I've tried modifying the requirements following the reference you provided, but some of the content is pretty opaque. In particular, what does run requirement `libblas * *openblas` mean? I can't find any documentation for this syntax either. In any event, the most recent build error message seems to be complaining about the absence of `libopenblas`.

Any further assistance would be much appreciated.

**Comment by ryanvolz** [comment]
Ok, I looked at the source package and the `setup.py`, and it appears that your problem is that this is not well-packaged on PyPI. The "source" already includes compiled libraries, and the "build" involves just copying those over, which explains why you didn't need the blas dependency to "build" but that it fails during testing.

Luckily, there is a part of the `setup.py` where you can set `USE_CYTHON=True` and it looks like it will do the actual build. So I think you'll need to patch `setup.py` so that `USE_CYTHON=True` and see what actually building it accomplishes.

Figuring out the blas stuff can happen once that's working.

**Comment by bwohlberg** [comment]
The unusual structure of the source package is unfortunately due to the authors' inability to release the source for the core algorithm implementation, so switching to building all the binaries is not an option (I suspect that the `USE_CYTHON` variable is purely for their internal usage). However, the linux package is also building successfully now, which just leaves the osx build to be fixed. My first thought on seeing the `Illegal instruction` error was there's some sort of binary incompatibility with the binaries in the source package, but a local build on the identical OS version passes without any issues. Do you have any idea what the problem might be, or what the best approach would be for debugging?

**Comment by ryanvolz** [comment]
Ok, then this gets into binary repackaging and that is outside of my experience with conda-forge. I'll have to call on @conda-forge/help-c-cpp for guidance.

Looking at the actual license, it looks like something custom and not BSD-4-Clause, so that will have to clarified as well. I think custom licenses are allowed (with the appropriate SPDX identifier), as long as they allow redistribution. I'm not sure if the restrictions in this particular license qualify, particularly the third clause:

    (iii) the software, or any reproduction or modification thereof, will not be copied or posted on any network computer or broadcast in any media without prior consent is obtained from TAU

Then assuming that conda-forge can still package this, I take back what I initially said about "noarch: python". I don't know how well it will work in practice for system compatibility (`Illegal instruction` on macOS is not a good sign), but since this just throws all of the compiled libs into the Python package, it's better off as noarch so that a bunch of identical but differently-named packages are not uploaded. Either that or you should write a "build" script to copy the platform-appropriate libraries into the correct places for each platform so that the packages are not all identical. You should be able to drop all "build" and "host" dependencies and only worry on getting the "run" dependencies correct.

If I had to guess, I think the "illegal instruction" error on macOS probably has something to do with the repackaged library linking in Python, and then when it is loaded with the conda-forge Python there is a conflict. [This](https://blog.tim-smith.us/2015/09/python-extension-modules-os-x/) describes the problem along these lines that I've encountered before.

**Comment by bwohlberg** [comment]
Thanks! I'll follow up on the license first since there doesn't seem to be much point in proceeding if that issue can't be resolved.

**Comment by bwohlberg** [comment]
@ryanvolz: The license identifier in the recipe has been fixed. I've also discussed the license clause that you quoted with the original package owner. I'm hoping that the language of this clause can be modified, but if not, explicit permission for conda-forge to redistribute the package can apparently been obtained.

 @conda-forge/help-c-cpp: Your help with the OSX test failure would be appreciated.

**Comment by bwohlberg** [comment]
@conda-forge/help-c-cpp: This is another request for help with the OSX test failure.

**Comment by bwohlberg** [comment]
@conda-forge/help-c-cpp: Your help with the OSX test failure would be appreciated.

**Review by carterbox — CHANGES_REQUESTED** [review]
Two problems in order of importance:

1. The license does not allow conda-forge to distribute this software without permission from Tampere University. I think we need the original author of the package to republish it with a license that allows either anyone to redistribute or conda-forge specifically to redistribute. IMO, the license doesn't even allow PYPI to distribute this package because PYPI isn't named in the license.

2. Revending pre-compiled binaries.

> The "source" already includes compiled libraries, and the "build" involves just copying those over

Revending pre-built binaries is not allowed on conda-forge; generally we require all binaries to be built with our compilers. It looks like the cython modules can be recompiled if you patch the setup to set `USE_CYTHON=True`.

Please ping the team again when you have addressed these issues.

**Comment by bwohlberg** [comment]
I believe the license issue is being worked on by the original package authors.

Is there any exception to the policy on revending of pre-build binaries? I expect that the cython interface could be compiled as you suggest, but the `pypi` package also contains pre-built libraries for the core algorithms because the authors cannot get authorization to release the source.

**Pull request closed without merging**